pkg/vault/contracts/BalancerContractRegistry.sol

```diff
@@ -0,0 +1,201 @@
+     * );
+     * replaceBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool', 0x9FC3da866e7DF3a1c57adE1a97c9f00a70f010c8);
+     *
+     * At that point,
+     * getBalancerContract(ContractType.POOL_FACTORY, '20241205-v3-weighted-pool') returns active=false,
+     * isActiveBalancerContract(ContractType.POOL_FACTORY, 0x201efd508c8DfE9DE1a13c2452863A78CB2a86Cc) returns false,
+     * getBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool') returns the v2 address (and active=true).
+     */
+
+    /// @inheritdoc IBalancerContractRegistry
+    function registerBalancerContract(
+        ContractType contractType,
+        string memory contractName,
+        address contractAddress
+    ) external authenticate {
+        if (contractAddress == address(0)) {
+            revert ZeroContractAddress();
+        }
+
+        if (bytes(contractName).length == 0) {
+            revert InvalidContractName();
+        }
+
+        bytes32 contractId = _getContractId(contractType, contractName);
+
+        if (_contractRegistry[contractId] != address(0)) {
+            revert ContractAlreadyRegistered(contractType, contractName);
+        }
+
+        // Edge case: Cannot register a new "alias" (i.e., type/name) for an address if it's already been deprecated.
+        ContractInfo memory info = _contractInfo[contractAddress];
+        if (info.isRegistered && info.isActive == false) {
+            revert ContractAlreadyDeprecated(contractAddress);
+        }
+
+        // Store the address in the registry, under the unique combination of type and name.
+        _contractRegistry[contractId] = contractAddress;
+
+        // Record the address as active. The `isActive` flag enables differentiating between unregistered and deprecated
+        // addresses.
+        _contractInfo[contractAddress] = ContractInfo({
+            contractType: contractType,
+            isRegistered: true,
+            isActive: true
+        });
+
+        emit BalancerContractRegistered(contractType, contractName, contractAddress);
+    }
+
+    /// @inheritdoc IBalancerContractRegistry
+    function deprecateBalancerContract(address contractAddress) external authenticate {
+        ContractInfo memory info = _contractInfo[contractAddress];
+
+        // Check that the address has been registered.
+        if (info.isRegistered == false) {
+            revert ContractNotRegistered();
+        }
+
+        // If it was registered, check that it has not already been deprecated.
+        if (info.isActive == false) {
+            revert ContractAlreadyDeprecated(contractAddress);
+        }
+
+        // Set active to false to indicate that it's now deprecated. This is currently a one-way operation, since
+        // deprecation is considered permanent. For instance, calling `disable` to deprecate a factory (preventing
+        // new pool creation) is permanent.
```

Comment:
hmm but this can still be overcome by calling `replace`. Am I missing something?